### PR TITLE
Improve handling of long-lived streaming RPCs.

### PR DESCRIPTION
**Pull request by SanjayVas**
This prevents said RPCs from overloading the Kingdom.

- - -
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697)

### CODE REVIEW

**Review by renjiezh — CHANGES_REQUESTED** [review]
> Reviewed 4 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697)* status: 4 of 8 files reviewed, 3 unresolved discussions (waiting on @matthewclegg and @SanjayVas)

___
*[`src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt` line 97 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697#-NCafnmF3di8Jb10SxlE:-NCafnmF3di8Jb10SxlF:by0ioji) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/dade6930a309f0e29838361a86bf15cd0dd68ceb/src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt#L97)):*
> ```Kotlin
>     // TODO(world-federation-of-advertisers/cross-media-measurement#695): Use gRPC service config
>     // rather than custom retry logic w/backoff.
>     var attemptNumber = 1
> ```

nit: Do we need a attempt limit to crash the process so it is easier to know streaming goes wrong?
___
*[`src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt` line 101 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697#-NCafVgI9BIOloROrWVE:-NCafVgI9BIOloROrWVF:byf7o0) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/dade6930a309f0e29838361a86bf15cd0dd68ceb/src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt#L101)):*
> ```Kotlin
>       try {
>         lastProcessedContinuationToken = syncStatuses(lastProcessedContinuationToken)
>       } catch (e: StreamingException) {
> ```

nit: Can we explicitly have logs that it is hit by a StreamingException.
___
*[`src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt` line 104 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697#-NCaidunANX6vUJde5Fk:-NCaidunANX6vUJde5Fl:b-rnmkm3) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/dade6930a309f0e29838361a86bf15cd0dd68ceb/src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt#L104)):*
> ```Kotlin
>         when (e.cause.status.code) {
>           Status.Code.UNAVAILABLE,
>           Status.Code.DEADLINE_EXCEEDED -> {
> ```

I wonder if backoff delay needed for DEADLINE_EXCEEDED.
If I understand correctly, when everything goes well, herald will receive DEADLINE_EXCEEDED every 10 min (duration depends on the server setting). This is not an error. It should not increase the attempts number which eventually leads herald sleep for too long.

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697)* status: 4 of 8 files reviewed, 3 unresolved discussions (waiting on @matthewclegg, @renjiezh, and @SanjayVas)

___
*[`src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt` line 97 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697#-NCafnmF3di8Jb10SxlE:-NCb0nXcEcd-9pqpKnTX:b-1cxfn) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/dade6930a309f0e29838361a86bf15cd0dd68ceb/src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt#L97)):*
<details><summary><i>Previously, renjiezh wrote…</i></summary><blockquote>

nit: Do we need a attempt limit to crash the process so it is easier to know streaming goes wrong?
</blockquote></details>

I thought about this, but crashing will just result in the Herald restarting with it's own CrashLoopBackoff. I'm not entirely sure if that's better. I suppose it allows one to configure a failure limit at the K8s level and have the job just stop running entirely.

If we throw at this limit, then we should also just throw when we hit unexpected error codes.
___
*[`src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt` line 104 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697#-NCaidunANX6vUJde5Fk:-NCb1-JrBy7w4bJ1nqGt:bsmpwvj) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/dade6930a309f0e29838361a86bf15cd0dd68ceb/src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt#L104)):*
<details><summary><i>Previously, renjiezh wrote…</i></summary><blockquote>

I wonder if backoff delay needed for DEADLINE_EXCEEDED.
If I understand correctly, when everything goes well, herald will receive DEADLINE_EXCEEDED every 10 min (duration depends on the server setting). This is not an error. It should not increase the attempts number which eventually leads herald sleep for too long.
</blockquote></details>

It won't receive a `DEADLINE_EXCEEDED` under normal operation. Instead, the server just completes the flow at 10 minutes.

It would hit `DEADLINE_EXCEEDED` if it hits an RPC deadline anywhere in the chain.  I have an upcoming PR that adds a default deadline for calls to the Kingdom internal server from the API servers.

**Review by renjiezh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697)* status: 4 of 8 files reviewed, 2 unresolved discussions (waiting on @matthewclegg and @SanjayVas)

___
*[`src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt` line 97 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697#-NCafnmF3di8Jb10SxlE:-NCb4I_O-wpyQK822Nm2:b-wlcqeg) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/dade6930a309f0e29838361a86bf15cd0dd68ceb/src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt#L97)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

I thought about this, but crashing will just result in the Herald restarting with it's own CrashLoopBackoff. I'm not entirely sure if that's better. I suppose it allows one to configure a failure limit at the K8s level and have the job just stop running entirely.

If we throw at this limit, then we should also just throw when we hit unexpected error codes.
</blockquote></details>

I was thinking the way to detect failure. If a permanent disconnection to Kingdom happens, a restart count or even CrashLoopBackoff can help us be aware, while keeping delaying could hide the failure.

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697)* status: 3 of 8 files reviewed, all discussions resolved (waiting on @matthewclegg and @renjiezh)

___
*[`src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt` line 97 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697#-NCafnmF3di8Jb10SxlE:-NCb5wWk9K1gVPzPQbze:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/dade6930a309f0e29838361a86bf15cd0dd68ceb/src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt#L97)):*
<details><summary><i>Previously, renjiezh wrote…</i></summary><blockquote>

I was thinking the way to detect failure. If a permanent disconnection to Kingdom happens, a restart count or even CrashLoopBackoff can help us be aware, while keeping delaying could hide the failure.
</blockquote></details>

Done.
___
*[`src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt` line 101 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697#-NCafVgI9BIOloROrWVE:-NCb5xMf9iUJm5D5tu8W:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/dade6930a309f0e29838361a86bf15cd0dd68ceb/src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt#L101)):*
<details><summary><i>Previously, renjiezh wrote…</i></summary><blockquote>

nit: Can we explicitly have logs that it is hit by a StreamingException.
</blockquote></details>

Done.

**Review by renjiezh — APPROVED** [review]
> Reviewed 2 of 8 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @matthewclegg)

**Review by matthewclegg — CHANGES_REQUESTED** [review]
> Reviewed 1 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697)* status: all files reviewed, 4 unresolved discussions (waiting on @SanjayVas)

___
*[`src/main/k8s/duchy.cue` line 99 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697#-NCfQgee1G_PtL9DRxtF:-NCfQgee1G_PtL9DRxtG:bq33qhr) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/65335a00a47dcc756e24f152863275ca5e164505/src/main/k8s/duchy.cue#L99)):*
> ```CUE
> 				_kingdom_system_api_target_flag,
> 				_kingdom_system_api_cert_host_flag,
> 				_debug_verbose_grpc_client_logging_flag,
> ```

Please add Emad as a reviewer.  We will need to make corresponding changes to the yaml file of our internal Duchy deployment.
___
*[`src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt` line 99 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697#-NCfRIQi22llLpvYJ3hw:-NCfRIQjBvuh9AGCDakU:b-3mjfe3) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/65335a00a47dcc756e24f152863275ca5e164505/src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt#L99)):*
> ```Kotlin
>     // TODO(world-federation-of-advertisers/cross-media-measurement#695): Use gRPC service config
>     // rather than custom retry logic w/backoff.
>     var attemptNumber = 1
> ```

Could you please explain why you are removing the pollingThrottler? 
___
*[`src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt` line 100 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697#-NCfSgjF2J_ZMkNXag00:-NCfSgjF2J_ZMkNXag01:b-3qnwhb) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/65335a00a47dcc756e24f152863275ca5e164505/src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt#L100)):*
> ```Kotlin
>     // rather than custom retry logic w/backoff.
>     var attemptNumber = 1
>     while (coroutineContext.isActive) {
> ```

Why would the context not be active?
___
*[`src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt` line 142 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697#-NCfTmIy8rBkaM-VpuSw:-NCfTmIy8rBkaM-VpuSx:b-nxirqn) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/65335a00a47dcc756e24f152863275ca5e164505/src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt#L142)):*
> ```Kotlin
>     var lastProcessedContinuationToken = continuationToken
>     coroutineScope {
>       systemComputationsClient
> ```

I think a cleaner design would be to create a wrapper for systemComputationsClient.streamActiveComputations() that catches and handles any exceptions that might occur.  What exceptions are we expecting to have to handle, and why?

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697)* status: all files reviewed, 4 unresolved discussions (waiting on @matthewclegg)

___
*[`src/main/k8s/duchy.cue` line 99 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697#-NCfQgee1G_PtL9DRxtF:-NCv2idICO7Yv3TXyedf:b-4mdb4q) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/65335a00a47dcc756e24f152863275ca5e164505/src/main/k8s/duchy.cue#L99)):*
<details><summary><i>Previously, matthewclegg (Matthew Clegg) wrote…</i></summary><blockquote>

Please add Emad as a reviewer.  We will need to make corresponding changes to the yaml file of our internal Duchy deployment.
</blockquote></details>

Pinged him offline as he's not yet in the GitHub org.
___
*[`src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt` line 99 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697#-NCfRIQi22llLpvYJ3hw:-NCv2yVH7A_RE3-ZpC4S:b-cor5yr) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/65335a00a47dcc756e24f152863275ca5e164505/src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt#L99)):*
<details><summary><i>Previously, matthewclegg (Matthew Clegg) wrote…</i></summary><blockquote>

Could you please explain why you are removing the pollingThrottler? 
</blockquote></details>

We don't need to poll on an interval, but rather reconnect the streaming RPC whenever it completes. We generally want to do this immediately, unless there's an error in which case we delay using a backoff.
___
*[`src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt` line 100 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697#-NCfSgjF2J_ZMkNXag00:-NCv2zSV1k_ufDaj4Gcm:b642jk) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/65335a00a47dcc756e24f152863275ca5e164505/src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt#L100)):*
<details><summary><i>Previously, matthewclegg (Matthew Clegg) wrote…</i></summary><blockquote>

Why would the context not be active?
</blockquote></details>

Coroutine cancellation. This is the equivalent of `while (true)`, but what one should use instead when there's a coroutine context (e.g. in a suspendable function).

See https://kotlinlang.org/docs/cancellation-and-timeouts.html#making-computation-code-cancellable
___
*[`src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt` line 142 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697#-NCfTmIy8rBkaM-VpuSw:-NCv5KZJ5-4IvAwzfhUW:b-2epmaf) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/65335a00a47dcc756e24f152863275ca5e164505/src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt#L142)):*
<details><summary><i>Previously, matthewclegg (Matthew Clegg) wrote…</i></summary><blockquote>

I think a cleaner design would be to create a wrapper for systemComputationsClient.streamActiveComputations() that catches and handles any exceptions that might occur.  What exceptions are we expecting to have to handle, and why?
</blockquote></details>

The only exception we should be getting under normal operation from any RPC should be `StatusException`, and it should be caught immediately rather than bubbled up in order to know which RPC resulted in that exception.

Basically every RPC call should have the equivalent of

```kotlin
try {
  client.methodCall()
catch (e: StatusException) {
  // Handle
}
```

This is the equivalent for streaming RPCs, as those return flows. #664 and #684 were intended to do this everywhere, but some cases were missed.

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697)* status: 6 of 8 files reviewed, 4 unresolved discussions (waiting on @matthewclegg and @renjiezh)

*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697#-ND5TZi24PYT6090cfAp:-ND5TZi24PYT6090cfAq:bdwqz5x) (no related file):*
@renjiezh you may want to take another look, as I tweaked some of the exception handling from #674 and noted you in a couple TODOs.

Basically:
1. Coroutine `CancellationException` should usually be bubbled up rather than caught to ensure coroutine cancellation works as expected. See https://kotlinlang.org/docs/cancellation-and-timeouts.html#cancellation-is-cooperative
1. In a similar vein, you generally want your "catch-all" blocks to use `Exception` rather than `Throwable`, so that you don't suppress serious JVM errors. These are usually unrecoverable and should crash the application. See https://docs.oracle.com/javase/8/docs/api/java/lang/Error.html
1. `StatusException` should be caught at the lowest level, i.e. the call site of the RPC. This is because how a given error should be handled depends on which method was being called. If it needs to be handled at a higher level, it should be wrapped in another exception type to make that explicit to calling code.

___

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697)* status: 6 of 8 files reviewed, 4 unresolved discussions (waiting on @matthewclegg and @renjiezh)

*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697#-ND5TZi24PYT6090cfAp:-ND5WPsNDrURQjQiHQP_:b-8hm32q) (no related file):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

@renjiezh you may want to take another look, as I tweaked some of the exception handling from #674 and noted you in a couple TODOs.

Basically:
1. Coroutine `CancellationException` should usually be bubbled up rather than caught to ensure coroutine cancellation works as expected. See https://kotlinlang.org/docs/cancellation-and-timeouts.html#cancellation-is-cooperative
1. In a similar vein, you generally want your "catch-all" blocks to use `Exception` rather than `Throwable`, so that you don't suppress serious JVM errors. These are usually unrecoverable and should crash the application. See https://docs.oracle.com/javase/8/docs/api/java/lang/Error.html
1. `StatusException` should be caught at the lowest level, i.e. the call site of the RPC. This is because how a given error should be handled depends on which method was being called. If it needs to be handled at a higher level, it should be wrapped in another exception type to make that explicit to calling code.
</blockquote></details>

4.  `UNKNOWN` is not a status code that are considered safe to retry. I didn't change this, as I don't know if it's part of the workaround for the race condition mentioned in #585.

    `RESOURCE_EXHAUSTED` should usually only be retried when you know that the resource is no longer exhausted (ideally there's a way to check this). Often it's better just to avoid retrying on this code and just returning an error.

___

**Review by renjiezh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697)* status: 6 of 8 files reviewed, 4 unresolved discussions (waiting on @matthewclegg and @renjiezh)

*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697#-ND5TZi24PYT6090cfAp:-ND5Xvlx9vAuXELC682z:bym4lrc) (no related file):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

4.  `UNKNOWN` is not a status code that are considered safe to retry. I didn't change this, as I don't know if it's part of the workaround for the race condition mentioned in #585.

    `RESOURCE_EXHAUSTED` should usually only be retried when you know that the resource is no longer exhausted (ideally there's a way to check this). Often it's better just to avoid retrying on this code and just returning an error.
</blockquote></details>

Thank you. I saw these TODOs you noted. They will be fixed while resolving the [issue 585](https://github.com/world-federation-of-advertisers/cross-media-measurement/issues/585). It is a P1 task on the plan.

___

**Review by matthewclegg — APPROVED** [review]
> Reviewed 1 of 8 files at r1, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697)* status: 7 of 8 files reviewed, 1 unresolved discussion (waiting on @renjiezh and @SanjayVas)

___
*[`src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt` line 142 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697#-NCfTmIy8rBkaM-VpuSw:-ND5c2Eu2J4xZXOypeXV:b-yymlow) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/65335a00a47dcc756e24f152863275ca5e164505/src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt#L142)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

The only exception we should be getting under normal operation from any RPC should be `StatusException`, and it should be caught immediately rather than bubbled up in order to know which RPC resulted in that exception.

Basically every RPC call should have the equivalent of

```kotlin
try {
  client.methodCall()
catch (e: StatusException) {
  // Handle
}
```

This is the equivalent for streaming RPCs, as those return flows. #664 and #684 were intended to do this everywhere, but some cases were missed.
</blockquote></details>

Thank you for the really helpful explanations!

**Review by stevenwarejones — COMMENTED** [review]
> Reviewed 4 of 8 files at r1, 1 of 3 files at r2, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697)* status: all files reviewed, 2 unresolved discussions (waiting on @SanjayVas)

___
*[`src/main/k8s/duchy.cue` line 116 at r6](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697#-ND9kuP89z4H7n8aaDb-:-ND9kuP89z4H7n8aaDb0:b-7q2yvv) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/7b73738a065b51269656c2d7d6f836e2f9c8149e/src/main/k8s/duchy.cue#L116)):*
> ```CUE
> 						_kingdom_system_api_target_flag,
> 						_kingdom_system_api_cert_host_flag,
> 						"--channel-shutdown-timeout=3s",
> ```

why remove this shutdown timeout?

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697)* status: all files reviewed, 2 unresolved discussions (waiting on @SanjayVas and @stevenwarejones)

___
*[`src/main/k8s/duchy.cue` line 116 at r6](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697#-ND9kuP89z4H7n8aaDb-:-ND9qCO22qcPJNCmg9N_:bjppm69) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/7b73738a065b51269656c2d7d6f836e2f9c8149e/src/main/k8s/duchy.cue#L116)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

why remove this shutdown timeout?
</blockquote></details>

3s is already the default, so no need to specify it here.

**Review by stevenwarejones — APPROVED** [review]
> Reviewed 1 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697)* status: all files reviewed, 1 unresolved discussion (waiting on @SanjayVas)

**Review by SanjayVas — COMMENTED** [review]
> Dismissed @matthewclegg from a discussion.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @SanjayVas)

___
*[`src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt` line 142 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/697#-NCfTmIy8rBkaM-VpuSw:-ND_DGvVDg7QadsCuieh:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/65335a00a47dcc756e24f152863275ca5e164505/src/main/kotlin/org/wfanet/measurement/duchy/daemon/herald/Herald.kt#L142)):*
<details><summary><i>Previously, matthewclegg (Matthew Clegg) wrote…</i></summary><blockquote>

Thank you for the really helpful explanations!
</blockquote></details>

Done.
